src/hapi_schema/db_funding.py

```diff
@@ -0,0 +1,92 @@
+"""Funding subcategory table and view."""
+
+from datetime import datetime
+
+from sqlalchemy import (
+    CheckConstraint,
+    DateTime,
+    ForeignKey,
+    String,
+    select,
+    text,
+)
+from sqlalchemy.orm import Mapped, mapped_column, relationship
+
+from hapi_schema.db_location import DBLocation
+from hapi_schema.utils.base import Base
+from hapi_schema.utils.view_params import ViewParams
+
+
+class DBFunding(Base):
+    __tablename__ = "funding"
+    __table_args__ = (
+        CheckConstraint(
+            "requirements_usd >= 0.0",
+            name="requirements_usd",
+        ),
+        CheckConstraint(
+            "funding_usd >= 0.0",
+            name="funding_usd",
+        ),
+        CheckConstraint(
+            "funding_pct >= 0.0",
+            name="funding_pct",
+        ),
+        CheckConstraint(
+            "(reference_period_end >= reference_period_start) OR (reference_period_start IS NULL)",
+            name="reference_period",
+        ),
+    )
+
+    resource_hdx_id = mapped_column(
+        ForeignKey("resource.hdx_id", onupdate="CASCADE", ondelete="CASCADE"),
+        nullable=False,
+    )
+
+    appeal_code: Mapped[str] = mapped_column(
+        String(32), primary_key=True, nullable=False
```

Comment:
You can probably remove nullable=False since it's a primary key, ditto for the primary keys below.

Unless you think it should always be explicit, but then we should probably update it in all other tables.